mypy/checker.py

```diff
@@ -2506,6 +2506,68 @@ def conditional_type_map(expr: Expression,
         return {}, {}
 
 
+def is_callable_type(type: Optional[Type]) -> bool:
+    """Takes in a type and returns if that type is callable, that is,
+    if `type()` is well-typed (ignoring arguments)."""
+    callable_types = (FunctionLike, AnyType, TypeType)
+    if any(isinstance(type, callable_type) for callable_type in callable_types):
+        return True
+
+    if isinstance(type, UnionType):
+        return all(is_callable_type(item) for item in type.items)
+
+    if isinstance(type, TypeVarType):
+        return not is_callable_type(type.erase_to_union_or_bound())
+
+    if isinstance(type, Instance):
+        method = type.type.get_method('__call__')
+        if method:
+            return is_callable_type(method.type)
+
+    return False
+
+
+def conditional_callable_type_map(expr: Expression,
+                                  current_type: Optional[Type],
+                                  ) -> Tuple[TypeMap, TypeMap]:
+    """Takes in an expression and the current type of the expression
+
+    Returns a 2-tuple: The first element is a map from the expression to
+    the restricted type if it must be callable. The second element is a
+    map from the expression to the type it would hold if it weren't
+    callable. Considers AnyType to be both callable and uncallable."""
+    if not current_type:
+        return {}, {}
+
+    if isinstance(current_type, AnyType):
+        return {}, {}
+
+    if isinstance(current_type, UnionType):
+        callables = []  # type: List[Type]
+        uncallables = []  # type: List[Type]
+        for item in current_type.items:
+
+            if isinstance(item, AnyType):
+                # We want Any to be considered both callable and uncallable
+                callables.append(item)
+                uncallables.append(item)
+
+            else:
+                if is_callable_type(item):
```

Comment:
There's a problem here: nested unions (e.g. `Union[int, Union[str, float]]`) aren't necessarily flattened, so if the inner union contains a callable and a non-callable, that item is binned with the uncallables -- which is wrong. Try e.g.
```python
U = Union[int, Union[str, Callable[[], int]]]
def f(a: U) -> int:
    if callable(a):
        reveal_type(a)
        return a()
    else:
        reveal_type(a)
        return a
```
vs.
```python
U = Union[int, str, Callable[[], int]]
```